File: src/main/java/com/gmail/borlandlp/minigamesdtools/gui/hotbar/api/HotbarApiInst.java
package com.gmail.borlandlp.minigamesdtools.gui.hotbar.api;

import com.gmail.borlandlp.minigamesdtools.APIComponent;
import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.Hotbar;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.HotbarListener;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.type.HeldHotbar;
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.utils.Leveling;
import net.minecraft.server.v1_12_R1.PacketPlayOutExperience;
import org.bukkit.Bukkit;
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.bukkit.inventory.ItemStack;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Hashtable;
import java.util.Map;

public class HotbarApiInst implements HotbarAPI, APIComponent {
    private Map<Player, Hotbar> players = new Hashtable<>();
    private HotbarListener hotbarListener;
    private BukkitTask task;

    @Override
    public void onLoad() {
        this.hotbarListener = new HotbarListener(this);
        Bukkit.getServer().getPluginManager().registerEvents(this.hotbarListener, MinigamesDTools.getInstance());

        final HotbarApiInst task = this;
        this.task = new BukkitRunnable() {
            public void run() {
                task.update();
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 0, 5);
    }

    @Override
    public void onUnload() {
        HandlerList.unregisterAll(this.hotbarListener);
        this.task.cancel();

        for (Player player : this.getAll().keySet()) {
            try {
                this.unbindHotbar(player);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    public void update() {
        for(Player player : new ArrayList<>(this.getAll().keySet())) {
            if(player != null && !player.isDead()) {
                Hotbar hotbar = this.getAll().get(player);
                try {
                    hotbar.update();
                    hotbar.draw();
                } catch (Exception e) {
                    e.printStackTrace();
                    this.unbindHotbar(player);
                    Debug.print(Debug.LEVEL.WARNING, "Player " + player.getName() + " was unbinded due to problems with his hotbar.");
                }
            }
        }
    }

    private void clearPlayerHotbar(Player p) {
        for (int i = 0; i < 9; i++) {
            p.getInventory().setItem(i, null);
        }
    }

    @Override
    public void bindHotbar(Hotbar hotbar, Player player) {
        Debug.print(Debug.LEVEL.NOTICE, "Bind hotbar for Player[name:" + player.getName() + "]");
        if(hotbar instanceof HeldHotbar) {
            player.getInventory().setHeldItemSlot(8);
        }

        this.players.put(player, hotbar);

        PacketPlayOutExperience packet = new PacketPlayOutExperience((float) 0, 0, 0);
        ((CraftPlayer) player).getHandle().playerConnection.sendPacket(packet);
    }

    @Override
    public void unbindHotbar(Player player) {
        if(!this.isBindedPlayer(player)) {
            return;
        }

        Debug.print(Debug.LEVEL.NOTICE, "Unbind hotbar for Player[name:" + player.getName() + "]");
        this.clearPlayerHotbar(player);
        this.players.remove(player);

        PacketPlayOutExperience packet = new PacketPlayOutExperience(player.getExp(), player.getTotalExperience(), player.getLevel());
        ((CraftPlayer) player).getHandle().playerConnection.sendPacket(packet);
    }

    @Override
    public boolean isBindedPlayer(Player player) {
        return this.players.containsKey(player);
    }

    @Override
    public Hotbar getHotbar(Player player) {
        return this.players.get(player);
    }

    @Override
    public Map<Player, Hotbar> getAll() {
        return new HashMap<>(this.players);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/gui/hotbar/api/HotbarApiInst.kt
package com.gmail.borlandlp.minigamesdtools.gui.hotbar.api

import com.gmail.borlandlp.minigamesdtools.APIComponent
import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.Hotbar
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.HotbarListener
import com.gmail.borlandlp.minigamesdtools.gui.hotbar.type.HeldHotbar
import net.minecraft.server.v1_12_R1.PacketPlayOutExperience
import org.bukkit.Bukkit
import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask
import java.util.*

class HotbarApiInst : HotbarAPI, APIComponent {
    private val players: MutableMap<Player, Hotbar> = Hashtable()
    private var hotbarListener: HotbarListener? = null
    private var task: BukkitTask? = null

    override fun onLoad() {
        hotbarListener = HotbarListener(this)
        Bukkit.getServer().pluginManager.registerEvents(hotbarListener, MinigamesDTools.instance)
        val task = this
        this.task = object : BukkitRunnable() {
            override fun run() {
                task.update()
            }
        }.runTaskTimer(MinigamesDTools.instance, 0, 5)
    }

    override fun onUnload() {
        HandlerList.unregisterAll(hotbarListener)
        task!!.cancel()
        for (player in all.keys) {
            try {
                unbindHotbar(player)
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
    }

    fun update() {
        mutableListOf(all.keys).forEach { hotbarsSet ->
            hotbarsSet.forEach { player ->
                if (player.isDead == true) return
                all[player]!!.update()
                all[player]!!.draw()
            }
        }
    }

    private fun clearPlayerHotbar(p: Player?) {
        for (i in 0..8) {
            p!!.inventory.setItem(i, null)
        }
    }

    override fun bindHotbar(hotbar: Hotbar, player: Player) {
        Debug.print(
            Debug.LEVEL.NOTICE,
            "Bind hotbar for Player[name:" + player.name + "]"
        )
        if (hotbar is HeldHotbar) {
            player.inventory.heldItemSlot = 8
        }
        players[player] = hotbar
        val packet = PacketPlayOutExperience(0.toFloat(), 0, 0)
        (player as CraftPlayer?)!!.handle.playerConnection.sendPacket(packet)
    }

    override fun unbindHotbar(player: Player) {
        if (!isBindedPlayer(player)) {
            return
        }
        Debug.print(
            Debug.LEVEL.NOTICE,
            "Unbind hotbar for Player[name:" + player.name + "]"
        )
        clearPlayerHotbar(player)
        players.remove(player)
        val packet =
            PacketPlayOutExperience(player.exp, player.totalExperience, player.level)
        (player as CraftPlayer?)!!.handle.playerConnection.sendPacket(packet)
    }

    override fun isBindedPlayer(player: Player): Boolean {
        return players.containsKey(player)
    }

    override fun getHotbar(player: Player): Hotbar {
        return players[player]!!
    }

    override val all: Map<Player, Hotbar>
        get() = HashMap(players)
}